### DISCUSSION TITLE
Migration guide for Nuxt 3

### DISCUSSION THREAD

**Discussion by danielroe** [nuxt/nuxt#15981] [Announcements]
With the new Nuxt 3 RC on the horizon (see [discussion](https://github.com/nuxt/framework/discussions/3447)) it is important to start building and updating a migration guide for people moving their apps from Nuxt 2 to Nuxt 3.

The current draft of the migration guide is located at https://nuxt.com/docs/migration/overview, and we would particularly value community support for improving the guide and the experience of migrating from Nuxt 2.

Particularly the following tasks:

* Reading the migration guide
* Testing following the steps in the migration guide
* Raising issues for any missing information, advice from your own experience or other improvements to the guide

---

[👉](https://emojipedia.org/backhand-index-pointing-right/) **[See the Nuxt 3 roadmap!](https://nuxt.com/docs/community/roadmap)**

**Comment by tobiasdiez** [comment]
The migration guide is really good and covered almost everything that was needed for a migration. A few things I found out after a bit of research, which might be good to add to the docs:
- What changes in `package.json` are necessary (from nuxt2 and nuxt2+bridge).
- Rename `static` to `public`.

**Reply by amitleuva1987** [reply]
@tobiasdiez , I have the same thoughts over migration guide. I don't see any guide on 'what to add in package.json' and 'what to remove from package.json' . I am still confused. Do you have any idea of what changes needs to be made in 'package.json'?

**Reply by tobiasdiez** [reply]
essentially, everything nuxt 2 and vue 2 related, and unsupported/obsolete modules need to be removed (e.g. look at the lock file and look for things that depend on nuxt 2).
Then follow the install guide for nuxt 3 to install the new packages.
You can have a look here https://github.com/JabRef/JabRefOnline/pull/1009/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519

Once you figured it out for your project, it would be nice if you could summarize your findings and update the migration guide.

**Reply by amitleuva1987** [reply]
@tobiasdiez , Thank you very much for your reply

**Reply by bmulholland** [reply]
There's still no instructions on what changes in package.json. I'd like to try to switch my app from Bridge to nuxt 3 and see what works and what breaks, but I don't know how to do that.

**Comment by L422Y** [comment]
One of the big things I ran into and was not expecting is the difference / lack of `store` directory functionality, there's not really any mention of that in the migration guide, and the `store` directory is missing completely from the v3 documentation.

**Reply by nathanchase** [reply]
Perhaps a brief description of the changes (no more mutations, no more mapState/mapActions helpers - all unnecessary) and linking to: https://pinia.vuejs.org/cookbook/migration-vuex.html 

Also, Pinia's docs needs to be updated to reflect that in Nuxt 3, you **_don't_** want to install [@nuxtjs/composition-api](https://composition-api.nuxtjs.org/).
https://pinia.vuejs.org/ssr/nuxt.html

**Reply by toniengelhardt** [reply]
The problem is if you migrate from a Nuxt 2 project, you probably want to first migrate to Vuex 4 with Options API and then, when you have refactored everything else, to Pinia.

**Reply by bmulholland** [reply]
Yeah smooth migration paths here are really necessary. For every additional thing that changes alongside the Nuxt upgrade, risk and effort increases dramatically. I worry that the lack of an upgrade path will hold the ecosystem back -- it's certainly preventing me from upgrading.

**Reply by vedmant** [reply]
Yeah, lack of documentation for options API in Nuxt 3 is holding me back from upgrading as well. I'm trying Nuxt 3 with options API on a small new projects and there so much questions that are not addressed in documentation. Specially how to migrate Vuex, plugins, middlewares.

**Reply by vedmant** [reply]
If someone needs to keep similar to Nuxt 2 Vuex functionality here is a module that supports nested stores (however doesn't support separate files for actions/mutations, etc) https://github.com/vedmant/nuxt3-vuex

**Comment by iBobik** [comment]
I am migration from Nuxt 2 to Nuxt 3 and I miss how to properly upgrade packages.json (what to add and what to remove).

**Reply by nathanchase** [reply]
Try [Bridge](https://v3.nuxtjs.org/bridge/overview) first and see what breaks. 

Then try [a minimal Nuxt 3 project starter](https://v3.nuxtjs.org/getting-started/quick-start#new-project) with `npx nuxi init nuxt-app` and move packages over one at a time to ensure that none of them break your build. It's ultimately less headache than just copying your Nuxt 2 package.json, updating `nuxt` to 3, and hoping it all works.

Incidentally, I've also found that Bridge has ultimately been less of the bottleneck/roadblock of migrating Nuxt 2 to Nuxt 3 than the plugin ecosystem lagging behind. For my experience, what doesn't work in Nuxt 3 also doesn't work in Bridge - unless disabling all the things Bridge offers (vite/nitro/etc.).

**Comment by shvchuk** [comment]
Is there a way to use both Pinia and Vuex while migrating from Nuxt 2 to Nuxt 3?

**Reply by nathanchase** [reply]
You can, but it's not recommended: https://pinia.vuejs.org/ssr/nuxt.html#using-pinia-alongside-vuex

**Reply by toniengelhardt** [reply]
Yes, it works, but you will not have access to the Nuxt instance anymore.

**Comment by nathanchase** [comment]
As you go through it, it would be nice to see the Nuxt 2 code side-by-side or stacked with the Nuxt 3 code when viewing on a desktop sized viewport, instead of having to toggle between tabs. Sometimes being able to visually see the differences between the two code blocks makes it instantly more understandable - much like a diff between commits.
Requires toggle:
![image](https://user-images.githubusercontent.com/5963574/179975559-b78c1673-534d-46c0-8a5d-3ca7ea2ee258.png)

Better, seeing both at once:
![image](https://user-images.githubusercontent.com/5963574/179975025-958d01e1-fbe0-4912-ad79-2d3f6867e838.png)

**Comment by nathanchase** [comment]
It might seem very basic, but there's no mention in this guide of how to access the context (`useNuxtApp`) and how it's changed from Nuxt 2 to Nuxt 3. 

A huge amount of the plugin ecosystem of Nuxt 2 relies on accessing those plugins generally with a `this.$auth` or `this.$device` or whatever the plugin is injecting, so it would be useful in this guide to know that anywhere you needed to access those globally available plugins/methods/properties, you need to do something like the following:
```js
const ctx = useNuxtApp();
const $device = ctx.$device;
```

**Reply by L422Y** [reply]
I've been using plugins that use `provide` which populates into `this`:

```javascript
export default defineNuxtPlugin(async (context) => {
    const _auth = ...
    if (process.client) {
        return {
            provide: {
                myAuth: _auth
            }
        }
    }
})
```

should create `this.$myAuth`

**Reply by nathanchase** [reply]
But you can't access `this` inside of `<script setup>` in a layout/page/component.

**Comment by nathanchase** [comment]
I recently had to go through [quite a bit of hoops to get Nitro to serve Nuxt 3 assets with compression for production](https://github.com/unjs/nitro/issues/69), so I'd suggest at least some portion of documentation devoted towards production deployment optimization. I'm certainly not the only developer looking for the best Lighthouse scores! :)

**Comment by nathanchase** [comment]
There should be a mention of the new default value of `activeClass` and `exactActiveClass` for [NuxtLink](https://v3.nuxtjs.org/api/components/nuxt-link/).
Nuxt 2 `nuxt-link-active` -> Nuxt 3 `router-link-active`
Nuxt 2 `nuxt-link-exact-active` -> Nuxt 3 `router-link-exact-active`

**Comment by elsiosanchez** [comment]
Greetings to all. i would like to add Nuxt 3 to my element-Plus and vue 3 project. 

is there a guide to get started ? 


Since I'm messed up with how to work with the routes. I was using the vue3 router before but I don't know if with nuxt 3 I have to use the ```<nuxt-link>.```

If you know a complete nuxt course it would be great.

**Comment by Jamiewarb** [comment]
Leaving a small note to say that the migration guide for plugins doesn't mention how to migrate the `context` variable we used to have in Nuxt 2, and that it's now `nuxtApp`. Small, but can be confusing for those migrating that don't yet know this

**Comment by iPrytz** [comment]
I miss some good explanation on how to replace nuxtServerInit

**Reply by iBobik** [reply]
I saw just recommendations to just use Pinia store, but it's examples are not focused on this use-case.

I need a way how to load data once for all pages on server init and not every time when I useStore.

Pinia could be useful if we can load data in a store init and all useStore calls will wait for it.

**Reply by TheAlexLichter** [reply]
You can use a server-side Nuxt plugin for that ☺️

**Reply by nestle49** [reply]
My workaround:

**nuxt.config.ts**
```
plugins: [
        { src: '~/plugins/init.server.ts' }, // nuxtServerInit equivalent; must be the first server plugin
}
```

**init.server.ts**
```
import { Context } from '@nuxt/types';
import { useGlobalStateStore } from '~/store/globalState';

const initServer: (context: Context) => void = async (context: Context) => {
    const { req, app, route } = context;
    const host = req.hostname;
    const globalStateStore = useGlobalStateStore();
    ...
};

export default initServer;
```

**Reply by Drider** [reply]
But there is one point. Keep in mind that this plugin will also run when the error page is loaded. As an option, you can add a check on which pages you do not need to download data.

**Comment by toniengelhardt** [comment]
One important thing to keep in mind is that devs will most likely not directly upgrade from `2` Nuxt 2 Options API + Vuex to `3` Nuxt 3 Composition API + Pinia, but rather to `2.5` Nuxt 3 Options API + Vuex 4 first and then gradually to Composition API and Pinia.

You will anyway hit a lot of road blocks and have to swap out and refactor most plugins and a lot of code, so keeping DX in mind (especially for large projects), I think the focus should be to get from `2` to `2.5` with as little refactoring as possible and then be able to migrate the last 50% step by step.

**Reply by vedmant** [reply]
In my case I'm trying to upgrade to Nuxt 3, I'll probably upgrade to Pinia, but I don't consider using composition API, as options API is fully supported by Vue 3 and in my opinion allows for better app structure and cleaner code. Main problem with Nuxt 3 I see is lack of documentation for options API usage.

**Reply by vedmant** [reply]
I added a simple plugin for supporting Vuex on Nuxt 3 https://github.com/vedmant/nuxt3-vuex

**Reply by toniengelhardt** [reply]
Biggest issue with Vuex 4 seems to be that the store is not available in plugins, middleware, etc.

**Reply by vedmant** [reply]
@toniengelhardt It's available https://github.com/vedmant/nuxt3-vuex/blob/master/src/plugin.js provides `$store`

**Reply by toniengelhardt** [reply]
For me `useStore()` is available, but always undefined outside of a component setup function.

**Reply by vedmant** [reply]
@toniengelhardt it should be available with `this.$store` in options API if you have nuxtApp available you'll have store available. like `useNuxtApp().$store`

**Reply by toniengelhardt** [reply]
This logs undefined or me:

```ts
export default defineNuxtPlugin(() => {
  const nuxtApp = useNuxtApp()
  console.log('XXX', nuxtApp.$store)
}
```

I register the store like this:

```ts
import { store } from '@/store/store'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.use(store)
  return {
    provide: {
      store,
    },
  }
})
```

But I also have this persistedstate plugin, not sure if that is the problem:

```ts
// Inspired by: https://github.com/toniengelhardt/vuex-persistedstate

import merge from 'deepmerge'
import localForage from 'localforage'
import type { Store } from 'vuex'

interface Options<State> {
  baseKey?: string
  modules?: string[]
  paths?: string[]
  skipMutations?: string[]
  serializer?: (value: any) => string
  deserializer?: (value: any) => any
  onRehydrated?: (store: Store<State>) => void
}

const defaultSerializer = JSON.stringify

const defaultDeserializer = (value: any) => {
  if (typeof value === 'string') {
    return JSON.parse(value)
  } else if (typeof value === 'object') {
    return value
  }
  return undefined
}

export function createPersistedState<State>(options: Options<State>): (store: Store<State>) => void {
  const storage = localForage
  const baseKey = options.baseKey || 'vuex'
  const modules = options.modules || []
  const paths = options.paths || []
  const excludedMutations = options.skipMutations || []
  const serializer = options.serializer || defaultSerializer
  const deserializer = options.deserializer || defaultDeserializer
  const onRehydrated = options.onRehydrated || (() => {})

  async function setItem(key: string, item: Partial<State>) {
    return await storage.setItem(key, serializer(item))
  }

  async function getItem(key: string) {
    const value = await storage.getItem(key)
    try {
      return deserializer(value)
    } catch (error) {}
    return undefined
  }

  async function rehydrateStore(store: Store<State>) {
    // Rehydrate store when it is initialized.
    console.time('[vuex] reyhdrate')
    const savedState = await getItem(baseKey) || {}
    for (const module of modules) {
      const item = await getItem(`${baseKey}-${module}`)
      if (item) {
        savedState[module] = item
      }
    }
    const mergedState: State = merge(store.state, savedState, {
      arrayMerge: (store, saved) => saved,
      clone: false,
    })
    store.replaceState(mergedState)
    console.timeEnd('[vuex] reyhdrate')
  }

  // Assert storage.
  // Not sure if and why this is needed...
  storage.setItem('@@', '1')
  storage.removeItem('@@')

  return (store: Store<State>) => {
    store.commit('updateState', { ready: false })
    rehydrateStore(store)
      .then(() => {
        // Execute hook after rehydration.
        onRehydrated(store)
        // Subscribe to mutations.
        store.subscribe((mutation, state) => {
          if (!excludedMutations.includes(mutation.type)) {
            // Determine module.
            const partials = mutation.type.split('/')
            const module = partials.length > 1 ? partials[0] : undefined
            if (module) {
              // Get state[module] and save every key that is in 'paths' to vuex-{module}.
              console.time(`[vuex] persist module ${module}`)
              const moduleState: any = state[module as keyof typeof state]
              const persistedModuleState: any = {}
              Object.keys(moduleState as Object).forEach((_key) => {
                if (paths.includes(`${module}.${_key}`)) {
                  persistedModuleState[_key] = moduleState[_key]
                }
              })
              setItem(`${baseKey}-${module}`, persistedModuleState)
              console.timeEnd(`[vuex] persist module ${module}`)
            } else {
              // Get everything that is not a module from the state and save it to vuex if the key is in 'paths'.
              console.time('[vuex] persist state')
              const persistedState: any = {}
              Object.keys(state as Object).forEach((_key) => {
                if (!modules.includes(_key) && paths.includes(_key)) {
                  persistedState[_key] = state[_key as keyof State]
                }
              })
              setItem(baseKey, persistedState)
              console.timeEnd('[vuex] persist state')
            }
          }
        })
        store.commit('updateState', { ready: true })
      })
  }
}
```

**Comment by yeonjulee1005** [comment]
I agree @toniengelhardt 
For large projects, creating a new project seems to make things cleaner than migrating from `nuxt2` ->`nuxt3`.

**Comment by MURHAF-ELMASRI** [comment]
> With the new Nuxt 3 RC on the horizon (see [discussion](https://github.com/nuxt/framework/discussions/3447)) it is important to start building and updating a migration guide for people moving their apps from Nuxt 2 to Nuxt 3.
> 
> The current draft of the migration guide is located at [v3.nuxtjs.org/docs/migration/overview](https://v3.nuxtjs.org/docs/migration/overview), and we would particularly value community support for improving the guide and the experience of migrating from Nuxt 2.
> 
> Particularly the following tasks:
> 
> * Reading the migration guide
> * Testing following the steps in the migration guide
> * Raising issues for any missing information, advice from your own experience or other improvements to the guide
> 
> [👉](https://emojipedia.org/backhand-index-pointing-right/) **[See the Nuxt 3 roadmap!](https://v3.nuxtjs.org/community/roadmap)**

the current link does not work. need updated to https://nuxt.com/docs/migration/overview/

**Reply by danielroe** [reply]
thanks! ✅

**Comment by vedmant** [comment]
Is there a way to avoid using Typescript completely in the Nuxt app? Also I prefer to keep all my imports explicit, for example in `nuxt.config.js` it's not clear where `defineNuxtConfig` should be imported from, my IDE cannot find `defineNuxtConfig` export.

**Reply by danielroe** [reply]
No, you do not need to use TypeScript at all. You can write pure JS.

And the import is from `nuxt/config`. You can turn off auto-imports with:

```js
import { defineNuxtConfig } from 'nuxt/config'

export default defineNuxtConfig({
  imports: {
    autoImport: false
  }
})
```

... and then import each auto-importable util directly from `#imports`.

**Comment by harlan-zw** [comment]
Just dropping this here in case it helps anyone, my Nuxt 3 migration guide: https://harlanzw.com/blog/nuxt-3-migration-cheatsheet

**Reply by vedmant** [reply]
@harlan-zw How do you migrate asyncData() with support of redirects and 404 error on backend?

**Comment by vedmant** [comment]
> I register the store like this:

Try to debug it, see when this plug-in is run, if store is defined there,
then another plugin. Maybe your other plug-in runs before the store plug-in.

**Reply by toniengelhardt** [reply]
Yes, that might be the reason. I will dig into it. Thanks for clarifying, it is already great to know that it should work and that it is somehow an issue with my setup 🙏🏽
